demo/test.c

```diff
@@ -1362,6 +1363,288 @@ static int test_mp_decr(void)
    return EXIT_FAILURE;
 }
 
+/*
+   Cannot test mp_exp(_d) without mp_n_root and vice versa.
+   So one of the two has to be tested from scratch.
+
+   Numbers generated by
+   for i in {1..10}
+   do
+     seed=$(head -c 10000 /dev/urandom | tr -dc '[:digit:]' | head -c 120);
+     echo $seed;
+     convertbase $seed 10  64;
+   done
+
+   (The program "convertbase" uses libtommath's to/from_radix functions)
+
+   Roots were precalculated with Pari/GP
+
+   default(realprecision,1000);
+   for(n=3,100,r = floor(a^(1/n));printf("\"" r "\", "))
+
+   All numbers as strings to simplifiy things, especially for the
+   low-mp branch.
+*/
+static int test_mp_n_root(void)
+{
+   mp_int a, c, r;
+   int e;
+   int i, j;
+
+   const char *input[] = {
+      /*
+       * 7695055368100873513971520489282333897557005032980617084180385070\
+       * 39530690836861445204045548919181252491201760744537260787 x
+       */
+      "4n9cbk886QtLQmofprid3l2Q0GD8Yv979Lh8BdZkFE8g2pDUUSMBET/+M/YFyVZ3mBp",
```

Comment:
Why are you writing the numbers in base64 and add them in decimal as a comment? Either write them only in base64 or just write them in decimal which I would prefer.

Reply:
> Why are you writing the numbers in base64 and add them in decimal as a comment? 

For the same reason I left the tracing `puts("DDD")` in it ;-)

I calculated them in Pari/gp and converted them later with an external program to shorten the code lines. It was just a list that I ticked off (the little "x" at the end of the decimal numbers is that tick), not meant as a translation.

> [...] or just write them in decimal which I would prefer.

I just don't like the overly long lines (not every compiler accepts string constants in multiple parts). OCD maybe?

Reply:
Yes, but you can also split the decimal constants?

Reply:
Wat? I didn't read. Not every compiler supports multiple string constants?! Those are unsupported. I hardly believe that tommath would compile with such a thing since it is quite common practice to stringize and concatenate in macros.

Reply:
> Not every compiler supports multiple string constants?

No, " multiple string constants " is not the right nomenclature, I mean things like this, taking `printf`  as an example
```
printf( "this shall mimic a %s long "
          "line, cut in %c\n", "very", '2');
```
I vividly remember all the curses I uttered, although it was a long time ago. Nevertheless:

> Those [compilers] are unsupported.

The documentation has no restrictions in that regard. It includes the old Borland C compiler and even MSVC is supported. We should set some strict limits, e.g (versions #s arbitrary): GCC > 3, clang > 3.5, MSVC > 2012, icc > 19 and so on (mingw, djgpp, &hellip;), document them and test them. The exact version #s are open for discussion, of course. But not here, I think it deservers its own thread, pardon, issue.